### PR TITLE
feat: various metrics (part 2)

### PR DESCRIPTION
**Pull request by pimeys**
moar

### CODE REVIEW

File: engine/crates/engine-v2/src/engine.rs [hunk 1]
```diff
@@ -324,6 +327,16 @@ impl<R: Runtime> Engine<R> {
                 tracing::debug!(target: GRAFBASE_TARGET, "{message}")
             }
 
+            if let Some(attributes) = operation_metrics_attributes {
+                for error in response.errors() {
+                    self.operation_metrics.increment_graphql_errors(GraphqlErrorAttributes {
+                        code: error.code.to_string(),
```
Comment (Finistere) [inline engine/crates/engine-v2/src/engine.rs:333]: Change this to a `&'static str`

File: engine/crates/engine-v2/src/engine.rs [hunk 2]
```diff
@@ -324,6 +327,16 @@ impl<R: Runtime> Engine<R> {
                 tracing::debug!(target: GRAFBASE_TARGET, "{message}")
             }
 
+            if let Some(attributes) = operation_metrics_attributes {
+                for error in response.errors() {
```
Comment (Finistere) [inline engine/crates/engine-v2/src/engine.rs:331]: should be distinct error codes as discussed

File: engine/crates/engine-v2/src/engine.rs [hunk 3]
```diff
@@ -435,28 +455,46 @@ impl<'ctx, R: Runtime> PreExecutionContext<'ctx, R> {
         };
 
         let mut status: GraphqlResponseStatus = GraphqlResponseStatus::Success;
+
         struct Sender<'a> {
             sender: mpsc::Sender<Response>,
             status: &'a mut GraphqlResponseStatus,
+            operation_name: Option<String>,
+            client: Option<Client>,
+            operation_metrics: GraphqlOperationMetrics,
```
Comment (Finistere) [inline engine/crates/engine-v2/src/engine.rs:464]: why clone? could just be a `&GraphqlOperationMetrics`

File: engine/crates/telemetry/src/metrics/operation.rs [hunk 4]
```diff
@@ -23,6 +24,12 @@ pub struct GraphqlOperationMetrics {
     query_preparation_latency: Histogram<u64>,
     batch_sizes: Histogram<u64>,
     request_body_sizes: Histogram<u64>,
+    graphql_errors: Counter<u64>,
+}
+
+#[derive(Clone)]
```
Comment (Finistere) [inline engine/crates/telemetry/src/metrics/operation.rs:30]: You can just remove the `Clone`, you don't need it for the response sender stuff in engine-v2, and `Clone` isn't even used in this repo. Don't think it is in api either.

File: engine/crates/engine-v2/src/engine.rs (hunk 3 above)
Comment (pimeys) [reply engine/crates/engine-v2/src/engine.rs:464]: Yeah, we could, if the async block starting from line 430 would not take the ownership of self...

Comment (Finistere) [reply engine/crates/engine-v2/src/engine.rs:464]: `self` is retrieved back as `ctx`, so you can just do `ctx.engine.operation_metrics`

**Review by Finistere — APPROVED** [review]

**Pull request merged**